### PR TITLE
Allow operators to create stream & memory connections in python and C++

### PR DESCRIPTION
**Pull request by slacrherbst**
This PR enables the use of operators (>>, << & ==) for connecting stream and memory masters and slaves. Each of these operators can be used to replace streamConnect, streamConnectBiDir and busConnect functions. Equivalent operators are also supported in C++.

Example streamConnect replacements:
````python
axiStream >> fifo >> compress >> fileWriter.getChannel(0)

# The reverse direction is also supported
fileWriter.getChannel(0) << compress << fifo << axiStream
````

Example streamConnectBiDir replacement:
````python
axiStream == srp
````

Example busConnect replacement:
````python
memoryMaster >> memorySlave

# The reverse direction is also supported
memorySlave << memoryMaster
````

These connection methods also support python classes which serve as containers for stream and memory devices, as long as accessor and operator override methods are provided. As an example the following methods are included with the PrbsPair class to enable use of the >>, << and == stream operators.

````python
    def _getStreamMaster(self):
        return self.PrbsRx._getStreamMaster()

    def _getStreamSlave(self):
        return self.PrbsTx._getStreamSlave()

    def __rshift__(self,other):
        pyrogue.streamConnect(self,other)
        return other

    def __lshift__(self,other):
        pyrogue.streamConnect(other,self)
        return other

    def __eq__(self,other):
        pyrogue.streamConnectBiDir(other,self)
````

The streamConnect and streamConnectBiDir are included as helper functions. These functions are required to support both native stream and memory classes as well as generic classes which contain stream and memory classes, using the _getStreamMaster, _getStreamSlave, _getMemoryMaster and _getMemorySlave to gain access to the contained objects.

When using the C++ versions of the operators, you need to de-reference the left hand side:

````c
*mast >> slave;

// or for chained connections
*(*axiStream >> fifo) >> slave;
````

### CODE REVIEW

File: docs/src/interfaces/memory/connecting.rst [hunk 1]
```diff
@@ -11,33 +11,22 @@ A memory master and slave are connected using the following commands in python:
    import pyrogue
 
    # Assume we are crated a memory tree with two masters, masterA and masterB
```
Comment (ruck314) [inline docs/src/interfaces/memory/connecting.rst]: Change `crated ` to `creating`

File: docs/src/interfaces/memory/connecting.rst [hunk 2]
```diff
@@ -11,33 +11,22 @@ A memory master and slave are connected using the following commands in python:
...
+   # Connections can also go in the reverse order
+   srpV3 << hubB << masterB
 
 The equivelent code in C++ is show below:
 
 .. code-block:: c
 
-   #include <rogue/Helpers.h>
-
    // Assume we are crated a memory tree with two masters, masterA and masterB
```
Comment (ruck314) [inline docs/src/interfaces/memory/connecting.rst]: Change `crated ` to `creating`

File: docs/src/interfaces/stream/connecting.rst [hunk 3]
```diff
@@ -38,15 +36,14 @@ it is passed to the primary slave:
...
 And in C++:
 
 .. code-block:: c
 
-   #include <rogue/Helpers.h>
-
-   rogueStreamConnect(myMaster, mySlave)
+   *myMaster >> mySlave;
 
 In some cases rogue entties can serve as both a stream Master and Slave. This is often the case when
```
Comment (ruck314) [inline docs/src/interfaces/stream/connecting.rst]: Change `entties` to `entities`

File: docs/src/interfaces/stream/debugStreams.rst [hunk 4]
```diff
@@ -6,7 +6,7 @@ Debugging Streams
 
 The base stream :ref:`interfaces_stream_Slave` class can server as a debug Frame receiver. This can
 be usefull when you want to display some or all of the data streams used in your designs. In most
-cases the debug Slave module will be used as a stream tap. In order to enable debugging the setDebug()
+cases the debug Slave module will be used as a secndary slave. In order to enable debugging the setDebug()
```
Comment (ruck314) [inline docs/src/interfaces/stream/debugStreams.rst]: Change `secndary` to `secondary`

File: docs/src/interfaces/stream/debugStreams.rst [hunk 5]
```diff
@@ -6,7 +6,7 @@ Debugging Streams
 
 The base stream :ref:`interfaces_stream_Slave` class can server as a debug Frame receiver. This can
 be usefull when you want to display some or all of the data streams used in your designs. In most
```
Comment (ruck314) [inline docs/src/interfaces/stream/debugStreams.rst]: Change `usefull` to `useful`

File: docs/src/interfaces/stream/usingFifo.rst [hunk 6]
```diff
@@ -122,11 +111,8 @@ Below is the equivalent code in C++
...
    # Connect the src and dst
-   rogueStreamConnect(src, dst);
-
-   # Add the Fifo as a stream tap
-   rogueStreamTap(src, fifo);
+   src >> dst;
 
-   # Connect the monitor to the FIfo output
-   rogueStreamConnect(fifo, mon)
+   # Add the Fifo as a second stream and on to the monitor (reverse order show as an example)
```
Comment (ruck314) [inline docs/src/interfaces/stream/usingFifo.rst]: maybe add a comment that the FIFO will assert back pressure on the mon or maybe change this example to a FIFO that will not back pressure on a data path (a lot of time you don't want monitors (like live displays) to back pressure the stream source

File: include/rogue/interfaces/memory/Slave.h [hunk 7]
```diff
@@ -176,6 +177,18 @@ namespace rogue {
                 * @param transaction Transaction pointer as TransactionPtr
```
Comment (ruck314) [inline include/rogue/interfaces/memory/Slave.h]: Change `transaction Transaction` to `Transaction` (delete repeated word)

File: include/rogue/interfaces/stream/Master.h [hunk 8]
```diff
@@ -39,7 +44,7 @@ namespace rogue {
           * interfaces to one or more stream slave objects. The frst stream Slave is used
```
Comment (ruck314) [inline include/rogue/interfaces/stream/Master.h]: Change `frst` to `first`

File: include/rogue/interfaces/stream/Master.h [hunk 9]
```diff
@@ -39,7 +44,7 @@ namespace rogue {
           * interfaces to one or more stream slave objects. The frst stream Slave is used 
           * to allocated new Frame objects and it the last Slave to receive frame data.
```
Comment (ruck314) [inline include/rogue/interfaces/stream/Master.h]: Change `objects and it the last Slave` to `objects, and it is the last Slave`

File: tests/test_memory.py [hunk 10]
```diff
@@ -72,7 +73,8 @@ def __init__(self):
 
         # Create a memory gateway
         self.ms = rogue.interfaces.memory.TcpServer("127.0.0.1",9080);
-        pr.busConnect(self.ms,self.sim)
+        #self.ms >> self.sim
```
Comment (ruck314) [inline tests/test_memory.py]: Do you want to remove the commented code?   Or add another part in the test code that test `>>` (other direction) of the bus connecting?

**Review by ruck314 — APPROVED** [review]

**Comment by bengineerd** [comment]
Is there no way around having to dereference the left side when working in C++?
```c
*master >> slave;
```
It feels like the kind of thing that will constantly trip people up.

Is it still the left side when assigning in the other direction?
```c
*slave << master;
```

**Comment by slacrherbst** [comment]
In order to avoid that we would have to subclass the smart pointers, and some of the oeprators such as == would generate additional problems. It is always the left side since it is the entity which is being "called".

**Pull request merged**